src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -71,22 +78,39 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader loader)
             string filePath = null,
             bool isGenerated = false)
         {
-            return new DocumentInfo(new DocumentAttributes(id, name, folders, sourceCodeKind, filePath, isGenerated), loader);
+            return Create(id, name, folders, sourceCodeKind, loader, filePath, isGenerated, documentServiceProvider: null);
+        }
+
+        internal static DocumentInfo Create(
+            DocumentId id,
+            string name,
+            IEnumerable<string> folders,
+            SourceCodeKind sourceCodeKind,
+            TextLoader loader,
+            string filePath,
+            bool isGenerated,
+            IDocumentServiceProvider documentServiceProvider)
+        {
+            return new DocumentInfo(new DocumentAttributes(id, name, folders, sourceCodeKind, filePath, isGenerated), loader, documentServiceProvider);
         }
 
         private DocumentInfo With(
             DocumentAttributes attributes = null,
-            Optional<TextLoader> loader = default)
+            Optional<TextLoader> loader = default,
+            IDocumentServiceProvider documentServiceProvider = null)
         {
             var newAttributes = attributes ?? Attributes;
             var newLoader = loader.HasValue ? loader.Value : TextLoader;
+            var newDocumentServiceProvider = documentServiceProvider ?? DocumentServiceProvider;
```

Comment:
this seems to violate our 'With' pattern.  Basically, there is no way for someone to say use the With to *remove* a documentServiceProvider.

Reply:
you can't. that's whole purpose of having Null Object pattern.

Reply:
it wasn't clear to me that you had a null-object pattern for these guys.  earlier in the review it looked like there could be null-fields of IDocumentServiceProvider.  I was going off of that.  If that's not allowed/possible here, that's great :)  But i would def like to see that applied in the entire review.  Thanks!